sdk/lib/drivers/sptilib/sptilib.c

```diff
@@ -0,0 +1,1037 @@
+        Status = STATUS_SUCCESS;
+
+    /* Update the output buffer size */
+    Apt->DataTransferLength = Srb->DataTransferLength;
+    if (IsDirectMemoryAccess)
+    {
+        Irp->IoStatus.Information = Apt->Length;
+    }
+    else
+    {
+        if ((Apt->AtaFlags & ATA_FLAGS_DATA_IN) && (AptData.BufferOffset != 0))
+            Irp->IoStatus.Information = AptData.BufferOffset + Apt->DataTransferLength;
+        else
+            Irp->IoStatus.Information = Apt->Length;
+    }
+
+    DPRINT("APT: Return %lx, DTL %lu, %lu bytes\n",
+           Status,
+           Apt->DataTransferLength,
+           (ULONG)Irp->IoStatus.Information);
+    DPRINT("APT: Return curr %02x:%02x:%02x:%02x:%02x:%02x:%02x\n",
+           TaskFile[8 + 0],
+           TaskFile[8 + 1],
+           TaskFile[8 + 2],
+           TaskFile[8 + 3],
+           TaskFile[8 + 4],
+           TaskFile[8 + 5],
+           TaskFile[8 + 6]);
+    if (Apt->AtaFlags & ATA_FLAGS_48BIT_COMMAND)
+    {
+        DPRINT("APT: Return prev %02x:%02x:%02x:%02x:%02x:%02x:%02x\n",
+               TaskFile[0],
+               TaskFile[1],
+               TaskFile[2],
+               TaskFile[3],
+               TaskFile[4],
+               TaskFile[5],
+               TaskFile[6]);
+    }
+
+Cleanup:
+    SptiFreeIrpContext(IrpContext);
+    return Status;
+}
+
+CODE_SEG("PAGE")
+NTSTATUS
+SptiHandleScsiPassthru(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _Inout_ PIRP Irp,
+    _In_ ULONG MaximumTransferLength,
+    _In_ ULONG MaximumPhysicalPages)
+{
+    PSCSI_PASS_THROUGH Spt = Irp->AssociatedIrp.SystemBuffer;
+    PIO_STACK_LOCATION IoStack;
+    BOOLEAN IsDirectMemoryAccess;
+    PASSTHROUGH_DATA SptData;
+    ULONG SenseInfoOffset;
+    PUCHAR Cdb;
+    NTSTATUS Status;
+    PVOID DataBuffer;
+    PPASSTHROUGH_IRP_CONTEXT IrpContext;
+    PSCSI_REQUEST_BLOCK Srb;
+
+    PAGED_CODE();
+
+    IoStack = IoGetCurrentIrpStackLocation(Irp);
+
+    ASSERT(GET_IOCTL(IoStack) == IOCTL_SCSI_PASS_THROUGH ||
+           GET_IOCTL(IoStack) == IOCTL_SCSI_PASS_THROUGH_DIRECT);
+
+    IsDirectMemoryAccess = (GET_IOCTL(IoStack) == IOCTL_SCSI_PASS_THROUGH_DIRECT);
+
+    Status = SptiInitializeSpt(DeviceObject,
+                               Irp,
+                               IoStack,
+                               IsDirectMemoryAccess,
+                               MaximumTransferLength,
+                               MaximumPhysicalPages,
+                               Spt,
+                               &SptData,
+                               &SenseInfoOffset,
+                               &Cdb,
+                               &DataBuffer);
+    if (!NT_SUCCESS(Status))
+        return Status;
+
+    DPRINT("SPT: Flags %u, DTL %lu, sense len %u\n",
+           Spt->DataIn,
+           Spt->DataTransferLength,
+           Spt->SenseInfoLength);
+    DPRINT("SPT: CDB %02x:%02x:%02x:%02x:%02x:%02x:%02x:%02x:%02x:%02x\n",
+           Cdb[0],
```

Comment:
Style nit: This is typically one instance of these cases, where it's actually preferable to write more than one single parameter on a line :)
(this also applies to other similar DPRINTs in this file)

Reply:
Done. Learned something new, I guess.